federation/pkg/federatedtypes/hpa.go

```diff
@@ -936,3 +944,60 @@ func (a *HpaAdapter) minReplicasIncreasable(obj pkgruntime.Object) bool {
 	}
 	return false
 }
+
+// updateClusterListOnTargetObject passes the necessary info to the target object,
+// so that the corresponding controller can act on that.
+// This is used because if an hpa is active on a federated object it is supposed
+// to control the replicas and presence/absence of target object from federated clusters.
+func (a *HpaAdapter) updateClusterListOnTargetObject(fedHpa *autoscalingv1.HorizontalPodAutoscaler, scheduleStatus map[string]*replicaNums) error {
+	if len(fedHpa.Spec.ScaleTargetRef.Kind) <= 0 || len(fedHpa.Spec.ScaleTargetRef.Name) <= 0 {
+		// nothing to do
+		return nil
+	}
+
+	names := []string{}
+	for clusterName, replicas := range scheduleStatus {
+		if replicas != nil {
+			names = append(names, clusterName)
+		}
+	}
+	clusterNames := hpautil.ClusterNames{Names: names}
+	qualifiedKind := extensionsinternal.Kind(fedHpa.Spec.ScaleTargetRef.Kind)
+	targetObj, err := getRuntimeObjectForKind(a.client, qualifiedKind, fedHpa.Namespace, fedHpa.Spec.ScaleTargetRef.Name)
+	if errors.IsNotFound(err) {
+		// Nothing to do; the target object does not exist in federation.
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+
+	updatedObj := hpautil.SetHpaTargetClusterList(targetObj, clusterNames)
+	_, err = updateRuntimeObjectForKind(a.client, qualifiedKind, fedHpa.Namespace, updatedObj)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func getRuntimeObjectForKind(c federationclientset.Interface, kind schema.GroupKind, ns, name string) (pkgruntime.Object, error) {
+	switch kind {
+	case extensionsinternal.Kind("ReplicaSet"):
+		return c.ExtensionsV1beta1().ReplicaSets(ns).Get(name, metav1.GetOptions{})
+	case extensionsinternal.Kind("Deployment"):
+		return c.ExtensionsV1beta1().Deployments(ns).Get(name, metav1.GetOptions{})
+	default:
+		return nil, fmt.Errorf("Unsupported federated kind targeted by hpa: %v", kind)
+	}
+}
+
+func updateRuntimeObjectForKind(c federationclientset.Interface, kind schema.GroupKind, ns string, obj pkgruntime.Object) (pkgruntime.Object, error) {
```

Comment:
Yes, I completely agree, but as mentioned I did not find a better way for this.
Also, I have used Deployment and ReplicaSet only in the target objects, because those are the two objects which are supported in control plane as of now. I have added comments on top of the functions for now, with a TODO to update the documentation when I do that.